### PR TITLE
Add optional.ExtractValue() to be able to handle Option type within interface{} value

### PR DESCRIPTION
**Pull request by 6543**
as title ... add an helper func for optional.Option type(s)

because simple type cast wont work if you not know the exact T type too ... (or want your func handle all Option Types ...)

### CODE REVIEW

File: modules/optional/option.go [hunk 1]
```diff
@@ -43,3 +45,29 @@ func (o Option[T]) ValueOrDefault(v T) T {
 	}
 	return v
 }
+
+// ExcractValue return value or nil and bool if object was an Optional
+// it should only be used if you already have to deal with interface{} values
+// and expect an Option type within it.
+func ExcractValue(obj any) (any, bool) {
```
Comment (KN4CK3R) [inline modules/optional/option.go]: What is `Excract`? Should this be `Extract`?

**Review by KN4CK3R — COMMENTED** [review]
What's the usecase of this?

**Comment by 6543** [comment]
usecase would be to make 
https://github.com/go-gitea/gitea/blob/487ac9bf6c239ce897f1a2f6c4321d6f1769a22f/routers/web/user/home.go#L335
optional.Option aware etc...

**Comment by 6543** [comment]
should I add this usecase in here too?

File: modules/optional/option.go (hunk 1 above)
Comment (6543) [inline modules/optional/option.go]:
```suggestion
// ExtractValue return value or nil and bool if object was an Optional
// it should only be used if you already have to deal with interface{} values
// and expect an Option type within it.
func ExtractValue(obj any) (any, bool) {
```

File: modules/optional/option_test.go [hunk 2]
```diff
@@ -57,3 +57,46 @@ func TestOption(t *testing.T) {
 	assert.True(t, opt3.Has())
 	assert.Equal(t, int(1), opt3.Value())
 }
+
+func TestExcractValue(t *testing.T) {
+	val, ok := optional.ExcractValue("aaaa")
+	assert.False(t, ok)
+	assert.Nil(t, val)
+
+	val, ok = optional.ExcractValue(optional.Some("aaaa"))
```
Comment (6543) [inline modules/optional/option_test.go]:
```suggestion
	val, ok = optional.ExtractValue(optional.Some("aaaa"))
```

File: modules/optional/option_test.go [hunk 3]
```diff
@@ -57,3 +57,46 @@ func TestOption(t *testing.T) {
 	assert.True(t, opt3.Has())
 	assert.Equal(t, int(1), opt3.Value())
 }
+
+func TestExcractValue(t *testing.T) {
```
Comment (6543) [inline modules/optional/option_test.go]:
```suggestion
func TestExtractValue(t *testing.T) {
```

File: modules/optional/option_test.go [hunk 4]
```diff
@@ -57,3 +57,46 @@ func TestOption(t *testing.T) {
...
+	val, ok := optional.ExcractValue("aaaa")
```
Comment (6543) [inline modules/optional/option_test.go]:
```suggestion
	val, ok := optional.ExtractValue("aaaa")
```

File: modules/optional/option_test.go [hunk 5]
```diff
@@ -57,3 +57,46 @@ func TestOption(t *testing.T) {
...
+	val, ok = optional.ExcractValue(optional.None[float64]())
+	assert.True(t, ok)
+	assert.Nil(t, val)
+
+	val, ok = optional.ExcractValue(&fakeHas{})
+	assert.False(t, ok)
+	assert.Nil(t, val)
+
+	wrongType := make(fakeHas2, 0, 1)
+	val, ok = optional.ExcractValue(wrongType)
```
Comment (6543) [inline modules/optional/option_test.go]:
```suggestion
	val, ok = optional.ExtractValue(optional.None[float64]())
	assert.True(t, ok)
	assert.Nil(t, val)

	val, ok = optional.ExtractValue(&fakeHas{})
	assert.False(t, ok)
	assert.Nil(t, val)

	wrongType := make(fakeHas2, 0, 1)
	val, ok = optional.ExtractValue(wrongType)
```

File: modules/optional/option_test.go [hunk 6]
```diff
@@ -57,3 +57,46 @@ func TestOption(t *testing.T) {
...
+	assert.False(t, ok)
+	assert.Nil(t, val)
+}
+
+func toPtr[T any](val T) *T {
+	return &val
+}
+
```
Comment (6543) [inline modules/optional/option_test.go]:
```suggestion

```

File: modules/optional/option.go [hunk 7]
```diff
@@ -43,3 +45,29 @@ func (o Option[T]) ValueOrDefault(v T) T {
 	}
 	return v
 }
+
+// ExtractValue return value or nil and bool if object was an Optional
+// it should only be used if you already have to deal with interface{} values
+// and expect an Option type within it.
+func ExtractValue(obj any) (any, bool) {
```
Comment (lunny) [inline modules/optional/option.go:52]: But the function has never been used?

Comment (6543) [reply modules/optional/option.go:52]: https://github.com/go-gitea/gitea/pull/29794#issuecomment-1997340942

Comment (lunny) [reply modules/optional/option.go:52]: I think yes. At least one example.

Comment (6543) [reply modules/optional/option.go:52]: ok :)

**Comment by 6543** [comment]
@lunny @KN4CK3R added an example usage

**Comment by KN4CK3R** [comment]
How/Where gets the `optional.Option` into `ctx.Data`?

**Comment by 6543** [comment]
if we start to adopt and use optionals more widly instead of type refs ...
e.g.
https://github.com/go-gitea/gitea/pull/29726/files#diff-aa8a6ecda61a1825f948895bec48df575bee830dd5c3d25fd81977452041945cR131-R133

**Comment by wxiaoguang** [comment]
Actually IMO the design of `AddParam` is totally wrong. It should add the key/value explicitly, but not depending on the "ctx.Data"

If the `AddParam` could be implemented clearly, then no need the `ExtractValue` trick.

**Comment by wxiaoguang** [comment]
->  Refactor AddParam to AddParamIfExist #29834

**Comment by 6543** [comment]
In this regards it looks like it is not jet needed, so I'll close this pull and **everybody is free to pick that patch** if needed in the future

**Pull request closed without merging**